File: 1/day1.go
package day1

import (
	"fmt"
	"os"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

// func run(filename string) int {
// 	fmt.Printf("open %s\n", filename)

// 	dat, err := os.ReadFile(filename)
// 	check(err)

// 	rows := strings.Split(string(dat), "\n")

// 	result := 0
// 	for _, value := range rows {
// 		first := 0
// 		second := 0

// 		for _, char := range value {
// 			// is_digit := char >= 48 && char <= 57
// 			is_digit := char >= '0' && char <= '9'

// 			if first == 0 && is_digit {
// 				first = int(char - '0')
// 			}

// 			if is_digit {
// 				second = int(char - '0')
// 			}
// 		}

// 		result += (first*10 + second)
// 	}

// 	return result
// }

var mapping map[string]int

func setup() {
	counts := [10]string{"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}

	mapping = make(map[string]int)
	for idx, name := range counts {
		mapping[name] = idx
	}
}

/*
def convert(text: str):
    for pointer in range(len(text)):
        subtext = text[pointer:]
        for word, number in mapping.items():
            if subtext.startswith(word):
                text = text[:pointer] + subtext.replace(word, str(number))
                break

    return text
*/

func convert(value string) string {
	for pointer := range value {
		if pointer > len(value) {
			// size of value was updated in loop, stop now!
			break
		}

		subtext := value[pointer:]

		for word, number := range mapping {
			if strings.HasPrefix(subtext, word) {
				value = value[:pointer] + strings.Replace(subtext, word, fmt.Sprint(number), 1)
				break
				// value = value[:pointer] + strings.Replace(subtext, value, fmt.Sprint(number), 1)
			}
		}
	}

	return value
}

func run(filename string) int {
	setup()
	fmt.Printf("open %s\n", filename)

	dat, err := os.ReadFile(filename)
	check(err)

	rows := strings.Split(string(dat), "\n")

	result := 0
	for _, value := range rows {
		first := 0
		second := 0

		value = convert(value)

		for _, char := range value {
			// is_digit := char >= 48 && char <= 57
			is_digit := char >= '0' && char <= '9'

			if first == 0 && is_digit {
				first = int(char - '0')
			}

			if is_digit {
				second = int(char - '0')
			}

		}

		fmt.Printf("%s - %d - %d\n", value, first, second)
		result += (first*10 + second)
	}

	return result
}

File: 1/day1.rs
use std::fs;
use std::collections::HashMap;

fn is_digit(char: char) -> bool {
    return char > '0' && char <= '9';
}

// fn process_line(line: &str) -> u32 {
//     let mut first = 0;
//     let mut last = 0;
//
//     for char in line.chars() {
//         if is_digit(char) {
//             last = char.to_digit(10).unwrap();
//
//             if first == 0 {
//                 first = char.to_digit(10).unwrap();
//             }
//         }
//     }
//
//     return first * 10 + last;
// }

fn get_numbers() -> HashMap<&'static str, usize> {
    let nums = vec!["zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];

    let mut result = HashMap::new();

    for (idx, &num) in nums.iter().enumerate() {
        result.insert(num, idx);
    }

    return result;
}

fn convert(line: &str) -> String {
    let nums = get_numbers();
    let mut modified_line = line.to_string();

    let mut i: usize = 0;
    for _ in 0..modified_line.chars().count() {
        if i > line.chars().count() {
            break;
        }

        let sliced = line.split_at(i).1;

        for num in nums.keys() {
            if sliced.starts_with(num) {
                let num_as_str = nums[num].to_string();

                // six -> 6ix
                let replacement = num_as_str + &num[1..];
                modified_line = modified_line.replacen(num, &replacement, 1);

                i += 1;
                // modified_line = modified_line.replacen(num, &num_as_str, 1);
                // i += num.chars().count() - 1;
                break;
            }
        }

        i += 1;
    }

    return modified_line;
}

fn process_line(line: &str) -> u32 {
    let mut first = 0;
    let mut last = 0;

    let updated = convert(line);

    for char in updated.chars() {
        if is_digit(char) {
            last = char.to_digit(10).unwrap();

            if first == 0 {
                first = char.to_digit(10).unwrap();
            }
        }
    }
    let result = first * 10 + last;
    return result;
}


fn run(filename: &str) -> u32 {
    let contents = fs::read_to_string(filename)
        .expect("Should have been able to read the file");

    let mut result = 0;
    for line in contents.split("\n") {
        println!("{}", line);
        let _result = process_line(&line);
        result += _result;
        println!("{}", _result);
    }

    return result;
}

fn main() {
    let result = run("dev2.txt");
    println!("result: {r}", r = result);
    assert!(result == 281);

    let pt2_result = run("test.txt");
    println!("part 2: {}", pt2_result);
}
